internal/fakecgo silently builds untested freebsd/386 and freebsd/arm (CGO_ENABLED=0)

> _Filed by Claude (Anthropic's Claude Code), on behalf of @hajimehoshi._

### PureGo Version

main @ 12ee364 (also present on earlier releases — the arch gate has been missing since `internal/fakecgo/go_freebsd.go` diverged from `go_netbsd.go`).

### Operating System

- [ ] Windows
- [ ] macOS
- [ ] Linux
- [X] FreeBSD
- [ ] NetBSD
- [ ] Android
- [ ] iOS

### Go Version (`go version`)

`go version go1.26.2 darwin/arm64` (host used to cross-compile; this is pure build-tag selection and is independent of the Go version).

### What steps will reproduce the problem?

In pure-Go mode (`CGO_ENABLED=0`, the `internal/fakecgo` path), cross-compile purego for 32-bit FreeBSD. It builds successfully, even though FreeBSD is only a supported tier on amd64/arm64:

```
$ CGO_ENABLED=0 GOOS=freebsd GOARCH=386 go build -gcflags="github.com/ebitengine/purego/internal/fakecgo=-std" .
$ echo $?
0
$ CGO_ENABLED=0 GOOS=freebsd GOARCH=arm go build -gcflags="github.com/ebitengine/purego/internal/fakecgo=-std" .
$ echo $?
0
```

The equivalent NetBSD targets correctly **refuse** to build:

```
$ CGO_ENABLED=0 GOOS=netbsd GOARCH=386 go build -gcflags="github.com/ebitengine/purego/internal/fakecgo=-std" .
# github.com/ebitengine/purego/internal/fakecgo
internal/fakecgo/callbacks.go:88:28: undefined: threadentry
internal/fakecgo/callbacks.go:89:28: undefined: x_cgo_init
internal/fakecgo/go_util.go:37:2: undefined: _cgo_sys_thread_start
```

This is purely build-tag selection and can be confirmed without any toolchain. On FreeBSD the bootstrap file `go_freebsd.go` is selected; on NetBSD `go_netbsd.go` is not:

```
$ CGO_ENABLED=0 GOOS=freebsd GOARCH=386 go list -f '{{.GoFiles}}' ./internal/fakecgo
[callbacks.go doc.go fakecgo.go freebsd.go go_freebsd.go go_libinit.go go_setenv.go go_util.go iscgo.go libcgo.go libcgo_freebsd.go setenv.go zsymbols.go zsymbols_freebsd.go]

$ CGO_ENABLED=0 GOOS=netbsd GOARCH=386 go list -f '{{.GoFiles}}' ./internal/fakecgo
[callbacks.go doc.go fakecgo.go go_libinit.go go_setenv.go go_util.go iscgo.go libcgo.go libcgo_netbsd.go netbsd.go setenv.go zsymbols.go zsymbols_netbsd.go]
```

### What is the expected result?

`freebsd/386` and `freebsd/arm` should refuse to build under `CGO_ENABLED=0`, exactly like `netbsd/386` and `netbsd/arm`. FreeBSD is documented as supported only on amd64/arm64 (README), is not built in CI, and is not run-tested — so the pure-Go path should not silently produce a 32-bit FreeBSD binary.

### What happens instead?

`internal/fakecgo/go_freebsd.go` is build-tagged just `//go:build !cgo`, with no architecture gate, so it compiles for every FreeBSD GOARCH. Its NetBSD counterpart is gated:

```
internal/fakecgo/go_freebsd.go:  //go:build !cgo
internal/fakecgo/go_netbsd.go:   //go:build !cgo && (amd64 || arm64)
```

The two files are byte-identical except for a NetBSD-specific signal-stack reset in `threadentry` (an OS-specific difference, not arch-specific). So the missing `(amd64 || arm64)` gate on FreeBSD is an oversight, not a deliberate 32-bit-FreeBSD capability — purego ships an untested pure-Go build for `freebsd/386` and `freebsd/arm`.

### Anything else you feel useful to add?

**Relationship to #460**

Same family as #460 — a build-tag-selection inconsistency in the cgo/fakecgo architecture matrix, surfaced by cross-compiling every `go tool dist list` target. It is the inverse direction: #460 was a cgo-fallback build that *should* have compiled but didn't; this is a pure-Go (fakecgo) build that *should not* exist but does.

**The cgo path is unaffected**

This is strictly about `CGO_ENABLED=0`. With `CGO_ENABLED=1` the entire `internal/fakecgo` package is excluded (every file is `!cgo`), and the main package's documented cgo fallback already supports `freebsd/386` — it selects `cgo.go` + `syscall_32bit.go`. So gating fakecgo removes only the untested pure-Go 32-bit build and leaves the integer-only cgo fallback intact, exactly how `netbsd/386` already behaves.

**Possible fix**

Gate only `internal/fakecgo/go_freebsd.go` to mirror `go_netbsd.go`:

```diff
-//go:build !cgo
+//go:build !cgo && (amd64 || arm64)
```

The other freebsd fakecgo files (`freebsd.go`, `libcgo_freebsd.go`, `zsymbols_freebsd.go`, `ztrampolines_freebsd.s`) stay plain `!cgo`, exactly as NetBSD gates only `go_netbsd.go`. After the change, `freebsd/386` and `freebsd/arm` fail to build with the same `undefined: threadentry` error as `netbsd/386`, while `freebsd/amd64` and `freebsd/arm64` still build and no other `go tool dist list` target changes.

--- a/internal/fakecgo/go_freebsd.go
+++ b/internal/fakecgo/go_freebsd.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !cgo
+//go:build !cgo && (amd64 || arm64)
 
 package fakecgo
 

